drv/psc-seq-server/src/main.rs

```diff
@@ -99,48 +99,198 @@
+/// timestamps in the main ringbuf.
+///
+/// An entry for each of the rectifier's PMBus status registers (e.g.
+/// `STATUS_WORD`, `STATUS_VOUT`, `STATUS_IOUT`, and so on...) is recorded read
+/// whenever a rectifier's `PWR_OK` pin changes state. Since exactly one of each
+/// register entry is recorded for every `Faulted` and `FaultCleared` entry, we
+/// don't really need to spend extra bytes on counting them, so they are marked
+/// as `count(skip)`.
+#[derive(Copy, Clone, PartialEq, Eq, counters::Count)]
+enum Trace {
+    #[count(skip)]
+    None,
+    Faulted {
+        now: u64,
+        #[count(children)]
+        psu: Slot,
+    },
+    FaultCleared {
+        now: u64,
+        #[count(children)]
+        psu: Slot,
+    },
+    StillInFault {
+        now: u64,
+        #[count(children)]
+        psu: Slot,
+    },
+    #[count(skip)]
+    StatusWord {
+        now: u64,
+        psu: Slot,
+        status_word: Result<u16, mwocp68::Error>,
+    },
+    #[count(skip)]
+    StatusIout {
+        now: u64,
+        psu: Slot,
+        status_iout: Result<u8, mwocp68::Error>,
+    },
+    #[count(skip)]
+    StatusVout {
+        now: u64,
+        psu: Slot,
+        status_vout: Result<u8, mwocp68::Error>,
+    },
+    #[count(skip)]
+    StatusInput {
+        now: u64,
+        psu: Slot,
+        status_input: Result<u8, mwocp68::Error>,
+    },
+    #[count(skip)]
+    StatusCml {
+        now: u64,
+        psu: Slot,
+        status_cml: Result<u8, mwocp68::Error>,
+    },
+    #[count(skip)]
+    StatusTemperature {
+        now: u64,
+        psu: Slot,
+        status_temperature: Result<u8, mwocp68::Error>,
+    },
+    #[count(skip)]
+    StatusMfrSpecific {
+        now: u64,
+        psu: Slot,
+        status_mfr_specific: Result<u8, mwocp68::Error>,
+    },
+    I2cError {
+        now: u64,
+        #[count(children)]
+        psu: Slot,
+        err: mwocp68::Error,
+    },
+    EreportSent {
+        now: u64,
+        #[count(children)]
+        psu: Slot,
+        class: ereport::Class,
+        len: usize,
+    },
+    EreportLost {
+        now: u64,
+        #[count(children)]
+        psu: Slot,
+        class: ereport::Class,
+        len: usize,
+        err: task_packrat_api::EreportWriteError,
+    },
+    EreportTooBig {
+        now: u64,
+        #[count(children)]
+        psu: Slot,
+        class: ereport::Class,
+    },
+}
+
+// Since entries in this ringbuffer contain timestamps, they will never be
+// de-duplicated. Thus, disable it.
+counted_ringbuf!(__TRACE, Trace, 32, Trace::None, no_dedup);
+
+/// PSU numbers represented as an enum. This is intended for use with
+/// `counted_ringbuf!`, instead of representing PSU numbers as raw u8s, which
+/// cannot derive `counters::Count` (and would have to generate a counter table
+/// with 256 entries rather than just 6).
+#[derive(Copy, Clone, Eq, PartialEq, counters::Count)]
+#[repr(u8)]
+enum Slot {
+    Psu0 = 0,
```

Comment:
Do we ever cast this to a `u8`?  If not, I'd suggest removing the explicit values.